File: python/routes/routesPost.py
import logging
import os
import base64
import cgi
import http.cookies
import jwt
import json
import bcrypt
import random
from http.server import BaseHTTPRequestHandler
from db.dbOperations import insert_user, select_user_by_name, insert_post, insert_post_image, insert_post_video, upvote, downvote, change_username, insert_avatar, update_post_image, update_post_video, insert_comment
from middleware.jwt import verify_jwt
import urllib.parse

# Configuração do logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

saltRounds = 10

def get_avatar_dir():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    base_dir = os.path.abspath(os.path.join(script_dir, '..', '..'))  
    return os.path.join(base_dir, 'python', 'public', 'img', 'users')

def generate_random_avatar():
    avatar_dir = get_avatar_dir()
    if not os.path.exists(avatar_dir):
        raise FileNotFoundError(f"Directory not found: {avatar_dir}")
    
    avatars = os.listdir(avatar_dir)
    if not avatars:
        raise FileNotFoundError("No avatars found in the directory.")
    
    selected_avatar = random.choice(avatars)
    with open(os.path.join(avatar_dir, selected_avatar), 'rb') as image_file:
        encoded_image = base64.b64encode(image_file.read()).decode('utf-8')
    return encoded_image

class routesPost(BaseHTTPRequestHandler):

    def do_POST(self):
        routes = {
            '/register': self.handle_register,
            '/login': self.handle_login,
            '/create-post': self.handle_create_post,
            '/upvote': self.handle_upvote,
            '/downvote': self.handle_downvote,
            '/create-comment': self.create_comment
        }

        if self.path in routes:
            content_type = self.headers.get('Content-Type')
            if content_type.startswith('application/json') or content_type.startswith('application/x-www-form-urlencoded') or content_type.startswith('multipart/form-data'):
                routes[self.path]()
            else:
                self.send_response(400)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': False, 'message': 'Content-Type must be application/json, application/x-www-form-urlencoded, or multipart/form-data'}).encode('utf-8'))
        else:
            self.handle_404()

    def handle_login(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length).decode('utf-8')
        user_data = dict(data.split('=') for data in post_data.split('&'))

        try:
            user = select_user_by_name(user_data['name'])
            if user and bcrypt.checkpw(user_data['password'].encode('utf-8'), user['password'].encode('utf-8')):
                token = jwt.encode({'name_user': user['name_user']}, os.getenv('JWT_SECRET'), algorithm='HS256')
                self.send_cookie_response(token)
            else:
                self.send_error_response(401, "Unauthorized: Incorrect username or password")
        except Exception as e:
            logger.error(f"Error during login: {str(e)}")
            self.send_error_response(500, f"Server error: {str(e)}")

    def send_cookie_response(self, token):
        cookie = http.cookies.SimpleCookie()
        cookie['jwt_token'] = token
        cookie['jwt_token']['path'] = '/'
        self.send_response(302)
        self.send_header('Location', '/home')
        self.send_header('Set-Cookie', cookie.output(header=''))
        self.end_headers()

    def handle_register(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length).decode('utf-8')
        user_data = dict(data.split('=') for data in post_data.split('&'))

        try:
            user_data['email'] = urllib.parse.unquote(user_data['email'])

            user = select_user_by_name(user_data['name'])
            if user:
                self.send_response(302)
                self.send_header('Location', '/register')
                self.end_headers()
                return

            hashed_password = bcrypt.hashpw(user_data['password'].encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
            
            # Inserir o usuário e obter o id do usuário inserido
            user_id = insert_user(user_data['name'], user_data['email'], hashed_password)

            # Gerar ou selecionar uma imagem aleatória para o avatar
            avatar_image = generate_random_avatar()

            # Inserir o avatar na tabela users_avatar
            insert_avatar(user_id, avatar_image)

            self.send_response(302)
            self.send_header('Location', '/login')
            self.end_headers()
        except Exception as e:
            logger.error(f"Error during registration: {str(e)}")
            self.send_error_response(500, f"Server error: {str(e)}")

    def handle_404(self):
        self.send_error_response(404, "Rota nao encontrada.")

    def send_error_response(self, code, message):
        self.send_error(code, message)

    @verify_jwt
    def handle_create_post(self):
        try:
            content_type = self.headers.get('Content-Type')
            content_length = int(self.headers['Content-Length'])

            if 'multipart/form-data' in content_type:
                fields = cgi.FieldStorage(fp=self.rfile, headers=self.headers, environ={'REQUEST_METHOD': 'POST', 'CONTENT_TYPE': content_type})
                post_content = {key: fields[key].value for key in fields if fields[key].value}

                post_title = post_content.get('title')  # Adicionando o título do post

                if 'file' in fields:
                    file_field = fields['file']
                    if self.is_valid_file_field(file_field):
                        file_data, file_name, file_size, file_type = self.read_file_field(file_field)
                        is_image, file_base64 = self.handle_file_upload(file_data, file_size, file_type)

                        user_name = self.decoded_token.get('name_user')
                        user = select_user_by_name(user_name)
                        if user:
                            if is_image:
                                insert_post_image(user['id_user'], post_title, post_content.get('content', ''), file_base64)
                            else:
                                insert_post_video(user['id_user'], post_title, post_content.get('content', ''), file_base64)

                            self.redirect_to('/home')
                        else:
                            self.send_error_response(404, "User not found")
                    else:
                        self.send_error_response(400, "File field is empty or invalid")
                else:
                    user_name = self.decoded_token.get('name_user')
                    user = select_user_by_name(user_name)
                    if user:
                        insert_post(user['id_user'], post_title, post_content.get('content', ''))
                        self.redirect_to('/home')
                    else:
                        self.send_error_response(404, "User not found")
            else:
                self.send_error_response(400, "Invalid content type")
        except Exception as e:
            logger.error(f"Exception during post creation: {e}")
            self.send_error_response(500, f"Server Error: {e}")

    
    def is_valid_file_field(self, file_field):
        return isinstance(file_field, cgi.FieldStorage) and hasattr(file_field, 'file') and hasattr(file_field, 'filename') and hasattr(file_field, 'type')

    def read_file_field(self, file_field):
        try:
            file_data = file_field.file.read()
            file_name = file_field.filename
            file_size = len(file_data)
            file_type = file_field.type
            return file_data, file_name, file_size, file_type
        except Exception as e:
            logger.error(f"Error reading file field: {e}")
            raise

    def handle_file_upload(self, file_data, file_size, file_type):
        image_types = ['image/gif', 'image/jpeg', 'image/png']
        video_types = ['video/mp4']
        max_image_size = 10 * 1024 * 1024  # 10 MB
        max_video_size = 10 * 1024 * 1024  # 10 MB

        if file_type in image_types and file_size <= max_image_size:
            return True, base64.b64encode(file_data).decode('utf-8')
        elif file_type in video_types and file_size <= max_video_size:
            return False, base64.b64encode(file_data).decode('utf-8')
        elif file_size == 0:
            return True, None
        else:
            raise ValueError("Invalid file type or size")

    def redirect_to(self, location):
        self.send_response(302)
        self.send_header('Location', location)
        self.end_headers()
        
    @verify_jwt
    def handle_upvote(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = json.loads(self.rfile.read(content_length).decode('utf-8'))
            post_id = post_data.get('post_id')

            user_name = self.decoded_token.get('name_user')
            user = select_user_by_name(user_name)
            if user:
                user_id = user['id_user']
                message = upvote(post_id, user_id)
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True, 'message': message}).encode('utf-8'))
            else:
                self.send_error_response(404, "User not found")
        except Exception as e:
            logger.error(f"Exception during upvote: {e}")
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'success': False, 'message': str(e)}).encode('utf-8'))


    @verify_jwt
    def handle_downvote(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = json.loads(self.rfile.read(content_length).decode('utf-8'))
            post_id = post_data.get('post_id')

            user_name = self.decoded_token.get('name_user')
            user = select_user_by_name(user_name)
            if user:
                user_id = user['id_user']
                message = downvote(post_id, user_id)
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True, 'message': message}).encode('utf-8'))
            else:
                self.send_error_response(404, "User not found")
        except Exception as e:
            logger.error(f"Error during downvote: {e}")
            self.send_error_response(500, "Server error during downvote")
        
    @verify_jwt
    def create_comment(self):
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length).decode('utf-8')
            comment_data = dict(data.split('=') for data in post_data.split('&'))

            user_name = self.decoded_token.get('name_user')
            user = select_user_by_name(user_name)
            if user:
                insert_comment(user['id_user'], comment_data['post_id'], comment_data['comment'])
                self.redirect_to('/home')
            else:
                self.send_error_response(404, "User not found")
        except Exception as e:
            logger.error(f"Error during comment creation: {e}")
            self.send_error_response(500, "Server error during comment creation")

    
            

File: nodejs/Routes/routesPost.js
const fs = require("fs");
const path = require("path");
const bcrypt = require("bcrypt");
const jwt = require("jsonwebtoken");
const db = require("../db/dbOperations");
const authenticateToken = require("../middleware/jwt");
const formidable = require("formidable");
require("dotenv").config();

const saltRounds = 10;

function sendCookieResponse(res, token, location) {
  res.writeHead(302, {
    "Set-Cookie": `jwt_token=${token}; Path=/; HttpOnly`,
    Location: location,
  });
  res.end();
}

async function handleRegister(req, res) {
  const form = new formidable.IncomingForm();
  form.parse(req, async (err, fields) => {
    if (err) {
      console.error(`Erro durante o parsing do formulário: ${err.message}`);
      res.writeHead(500, { "Content-Type": "text/plain" });
      res.end("Erro no servidor.");
      return;
    }

    const name = fields.name[0];
    const email = fields.email[0];
    const password = fields.password[0];

    try {
      const userExists = await db.selectUserByName(name);
      const emailExists = await db.selectUserByEmail(email);  
      if (userExists.length > 0) {
        const errorMessage = encodeURIComponent("O nome de usuário já está sendo utilizado");
        res.writeHead(302, { Location: `/register?error=${errorMessage}` });
        res.end();
        return;
      }

      if (emailExists.length > 0) {
        const errorMessage = encodeURIComponent("O email já está sendo utilizado");
        res.writeHead(302, { Location: `/register?error=${errorMessage}` });
        res.end();
        return;
      }

      const hashedPassword = await bcrypt.hash(password, saltRounds);
      const userId = await db.insertUser(name, email, hashedPassword);

      const avatarImage = generateRandomAvatar();
      await db.insertAvatar(userId, avatarImage);

      res.writeHead(302, { Location: "/login" });
      res.end();
    } catch (err) {
      console.error(`Erro durante o registro: ${err.message}`);
      res.writeHead(500, { "Content-Type": "text/plain" });
      res.end("Erro no servidor.");
    }
  });
}

async function handleLogin(req, res) {
  const form = new formidable.IncomingForm();
  form.parse(req, async (err, fields) => {
    if (err) {
      console.error(`Erro durante o parsing do formulário: ${err.message}`);
      res.writeHead(500, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ success: false, message: "Erro no servidor." }));
      return;
    }

    const name = fields.name[0];
    const password = fields.password[0];

    try {
      const user = await db.selectUserByName(name);

      if (user.length > 0 && bcrypt.compareSync(password, user[0].password)) {
        const token = jwt.sign(
          { name_user: user[0].name_user },
          process.env.JWT_SECRET,
          { algorithm: "HS256" }
        );
        sendCookieResponse(res, token, "/home");
      } else {
        const errorMessage = encodeURIComponent("Nome de usuário ou senha incorretos");
        res.writeHead(302, { Location: `/login?error=${errorMessage}` });
        res.end();
        return;
      }
    } catch (err) {
      console.error(`Erro durante o login: ${err.message}`);
      res.writeHead(500, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ success: false, message: "Erro no servidor." }));
    }
  });
}

function generateRandomAvatar() {
  const avatarDir = path.join(__dirname, "..", "public", "img");
  if (!fs.existsSync(avatarDir)) {
    throw new Error("O diretório de avatares não existe.");
  }
  const allFiles = fs.readdirSync(avatarDir);
  const avatars = allFiles.filter((file) =>
    [".jpg", ".jpeg"].includes(path.extname(file).toLowerCase())
  );
  if (avatars.length === 0) {
    throw new Error("Nenhum avatar JPG ou JPEG encontrado no diretório.");
  }
  const selectedAvatar = avatars[Math.floor(Math.random() * avatars.length)];
  const imagePath = path.join(avatarDir, selectedAvatar);
  const imageBuffer = fs.readFileSync(imagePath);
  return Buffer.from(imageBuffer).toString("base64");
}

function isValidFileField(file) {
  return file && file[0].size > 0 && file[0].mimetype;
}

function handleFileUpload(file) {
  const imageTypes = ["image/gif", "image/jpeg", "image/png"];
  const videoTypes = ["video/mp4"];
  const maxFileSize = 10 * 1024 * 1024;

  const fileData = fs.readFileSync(file[0].filepath);
  const fileSize = file[0].size;
  const fileType = file[0].mimetype;

  if (imageTypes.includes(fileType) && fileSize <= maxFileSize) {
    return { isImage: true, fileBase64: fileData.toString("base64") };
  } else if (videoTypes.includes(fileType) && fileSize <= maxFileSize) {
    return { isImage: false, fileBase64: fileData.toString("base64") };
  } else if (fileSize === 0) {
    throw new Error("Arquivo está vazio");
  } else {
    throw new Error("Tipo de arquivo ou tamanho inválido");
  }
}

async function handleCreatePost(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm({
      allowEmptyFiles: true,
      minFileSize: 0,
    });

    form.parse(req, async (err, fields, files) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("Erro no servidor.");
        return;
      }

      const title = fields.title ? fields.title[0] : null;
      
      const notcontent = " "; 
      const hascontent = fields.content[0]
      const content = hascontent ? hascontent : notcontent;  

      const file = files.file

      try {
        const userName = req.user.name_user;
        const user = await db.selectUserByName(userName);

        if (!user || user.length === 0) {
          res.writeHead(404, { "Content-Type": "text/plain" });
          res.end("Usuário não encontrado");
          return;
        }

        const userID = user[0].id_user;

        if (file && isValidFileField(file)) {
          const { isImage, fileBase64 } = handleFileUpload(file);

          if (isImage) {
            await db.insertPostWithImage(userID, title, content, fileBase64);
          } else {
            await db.insertPostWithVideo(userID, title, content, fileBase64);
          }
        } else {
          await db.insertPost(userID, title, content);
        }

        res.writeHead(302, { Location: "/home" });
        res.end();
      } catch (err) {
        console.error(`Erro durante a criação do post: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end(`Erro no servidor: ${err.message}`);
      }
    });
  });
}

async function handleUpVote(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm();
    form.parse(req, async (err, fields) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("Erro no servidor.");
        return;
      }

      const postId = fields.post_id;
      const userName = req.user.name_user;

      try {
        const user = await db.selectUserByName(userName);
        if (!user || user.length === 0) {
          res.writeHead(404, { "Content-Type": "text/plain" });
          res.end("Usuário não encontrado");
          return;
        }

        const userId = user[0].id_user;
        if (userId > 0) {
          await db.upvote(postId, userId);
        }

        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({ success: true, message: "Upvoted successfully" })
        );
      } catch (err) {
        console.error(`Erro durante a votação: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end(`Erro no servidor: ${err.message}`);
      }
    });
  });
}

async function handleDownVote(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm();
    form.parse(req, async (err, fields) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("Erro no servidor.");
        return;
      }

      const postId = fields.post_id;
      const userName = req.user.name_user;

      try {
        const user = await db.selectUserByName(userName);
        if (!user || user.length === 0) {
          res.writeHead(404, { "Content-Type": "text/plain" });
          res.end("Usuário não encontrado");
          return;
        }

        const userId = user[0].id_user;
        if (userId > 0) {
          await db.downvote(postId, userId);
        }

        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({ success: true, message: "Downvoted successfully" })
        );
      } catch (err) {
        console.error(`Erro durante a votação: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end(`Erro no servidor: ${err.message}`);
      }
    });
  });
}

async function handleCreateComment(req, res) {
  authenticateToken(req, res, () => {
    const form = new formidable.IncomingForm();
    form.parse(req, async (err, fields) => {
      if (err) {
        console.error(`Erro durante o parsing do formulário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("Erro no servidor.");
        return;
      }

      const postId = fields.post_id ? fields.post_id[0] : null;
      const comment = fields.comment ? fields.comment[0] : null;
      const userName = req.user.name_user;

      try {
        const user = await db.selectUserByName(userName);
        if (!user || user.length === 0) {
          res.writeHead(404, { "Content-Type": "text/plain" });
          res.end("Usuário não encontrado");
          return;
        }

        const userId = user[0].id_user;
        await db.insertComment(userId, postId, comment);

        res.writeHead(302, { Location: `/comments?post_id=${postId}` });
        res.end();
      } catch (err) {
        console.error(`Erro durante a criação do comentário: ${err.message}`);
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("Erro no servidor.");
      }
    });
  });
}

function handleLogout(req, res) {
  res.writeHead(302, {
    "Set-Cookie":
      "jwt_token=; Path=/; HttpOnly; Expires=Thu, 01 Jan 1970 00:00:00 GMT",
    Location: "/login",
  });
  res.end();
}

function setupPostRoutes(req, res) {
  const parsedUrl = require("url").parse(req.url, true);
  const pathName = parsedUrl.pathname;

  const postRoutes = {
    "/register": handleRegister,
    "/login": handleLogin,
    "/logout": handleLogout,
    "/create-post": handleCreatePost,
    "/upvote": handleUpVote,
    "/downvote": handleDownVote,
    "/create-comment": handleCreateComment,
  };

  if (req.method === "POST" && postRoutes[pathName]) {
    postRoutes[pathName](req, res);
  } else if (req.method === "POST") {
    res.writeHead(404, { "Content-Type": "text/plain" });
    res.end("Not Found");
  } else {
    res.writeHead(405, { "Content-Type": "text/plain" });
    res.end("Method Not Allowed");
  }
}

module.exports = setupPostRoutes;
